File: zsystem_customize/public/js/item.js
frappe.ui.form.on('Item Barcode', {
    barcode: function(frm, cdt, cdn) {
        let row = locals[cdt][cdn];

        if (row.barcode) {
            // Split the string by '|'
            let parts = row.barcode.split('|');

            // Find the part that starts with '1P' (case-insensitive)
            let target = parts.find(p => /^1P/i.test(p));

            if (target) {
                // Remove '1P' from start, dashes, and pipes
                let cleaned = target.replace(/^1P\s*/, '').replace(/-/g, '');

                // Add back the 'P' if needed
                cleaned = cleaned;

                // Update the field
                frappe.model.set_value(cdt, cdn, 'barcode', cleaned);
            }
        }
    }    
});

frappe.ui.form.on("Item", {
    custom_warehouse(frm) {
        if (!frm.doc.custom_warehouse) return;

        if (!frm.doc.item_defaults?.length) {
            let row = frm.add_child("item_defaults");
            row.default_warehouse = frm.doc.custom_warehouse;
        } else {
            frm.doc.item_defaults.forEach(row => {
                row.default_warehouse = frm.doc.custom_warehouse;
            });
        }

        frm.refresh_field("item_defaults");
    }
});

File: zsystem_customize/item.py
import frappe

def set_default_warehouse(doc, method):
    if not doc.custom_warehouse:
        return

    company = frappe.defaults.get_global_default("company")

    existing = next(
        (d for d in doc.item_defaults if d.company == company),
        None
    )

    if existing:
        existing.default_warehouse = doc.custom_warehouse
    else:
        doc.append("item_defaults", {
            "company": company,
            "default_warehouse": doc.custom_warehouse
        })